pkg/monitoring/monitoring.go

```diff
@@ -0,0 +1,169 @@
+ * under the terms of the Snowplow Limited Use License Agreement, Version 1.1
+ * located at https://docs.snowplow.io/limited-use-license-1.1
+ * BY INSTALLING, DOWNLOADING, ACCESSING, USING OR DISTRIBUTING ANY PORTION
+ * OF THE SOFTWARE, YOU AGREE TO THE TERMS OF SUCH LICENSE AGREEMENT.
+ */
+
+package monitoring
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+type MonitoringEvent struct {
+	Schema string         `json:"schema"`
+	Data   MonitoringData `json:"data"`
+}
+
+type MonitoringData struct {
+	AppName    string            `json:"appName"`
+	AppVersion string            `json:"appVersion"`
+	Tags       map[string]string `json:"tags"`
+
+	Message string `json:"message,omitempty"`
+}
+
+// MonitoringSender describes the interface for how to send heartbeat & alert events
+type MonitoringSender interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// Monitoring holds a new client and supporting data for sending heartbeat & alert events
+type Monitoring struct {
+	appName           string
+	appVersion        string
+	client            MonitoringSender
+	endpoint          string
+	tags              map[string]string
+	heartbeatInterval time.Duration
+	alertChan         chan error
+	log               *logrus.Entry
+	alertInterval     time.Duration
+
+	exitSignal chan struct{}
+}
+
+func NewMonitoring(appName, appVersion string, client MonitoringSender, endpoint string, tags map[string]string, heartbeatInterval, alertInterval time.Duration, alertChan chan error) *Monitoring {
+	return &Monitoring{
+		appName:           appName,
+		appVersion:        appVersion,
+		client:            client,
+		endpoint:          endpoint,
+		tags:              tags,
+		heartbeatInterval: heartbeatInterval,
+		log:               logrus.WithFields(logrus.Fields{"name": "Monitoring"}),
+		alertChan:         alertChan,
+		alertInterval:     alertInterval,
+		exitSignal:        make(chan struct{}),
+	}
+}
+
+func (m *Monitoring) Start() {
+	var once sync.Once
+
+	header := http.Header{}
+	header.Add("Content-Type", "application/json")
+
+	ticker := time.NewTicker(m.heartbeatInterval)
+
+	go func() {
+
+		alertCooldown := time.NewTicker(time.Millisecond * 100)
+	OuterLoop:
+		for {
+			select {
+			case <-ticker.C:
+				m.log.Info("Sending heartbeat")
+				if m.client != nil {
+					event := MonitoringEvent{
+						Schema: "iglu:com.snowplowanalytics.monitoring.loader/heartbeat/jsonschema/1-0-0",
+						Data: MonitoringData{
+							AppName:    m.appName,
+							AppVersion: m.appVersion,
+							Tags:       m.tags,
+						},
+					}
+					req, err := m.prepareRequest(event)
+					if err != nil {
+						m.log.Warnf("failed to prepare heartbeat event request: %s", err)
+						continue
+					}
+
+					req.Header = header
+					_, err = m.client.Do(req)
+					if err != nil {
+						m.log.Warnf("failed to send heartbeat event: %s", err)
+					}
+				}
+			case <-alertCooldown.C:
```

Comment:
I think I see what you're doing here, but I'm not sure I fully grasp the approach/what I may have missed.

Would we perhaps have a simpler implementation with something along these lines?

```
case <- m.alertChan:

if tstamp < (lastAlertTstamp + someInterval) {
continue
}

// sendAlert and set lastAlertTstamp
```